src/server/auth/server/api_server.go

```diff
@@ -698,11 +741,6 @@ func (a *apiServer) ModifyRoleBindingInTransaction(
 		return nil, err
 	}
 
-	roles, err := rolesFromRoleSlice(req.Roles)
-	if err != nil {
-		return nil, err
-	}
-
 	if strings.HasPrefix(req.Principal, auth.PachPrefix) && req.Resource.Type == auth.ResourceType_CLUSTER {
```

Comment:
Proposal: something I've been thinking about with this change is that with the old 3-role system, if you were a READER on a repo, the only way to add a pipeline to that repo is if you also had access to all of the pipeline's other inputs & output (and therefore could update the pipeline—i.e. you couldn't directly add `pipeline:foo` to one ACL unless you were an OWNER on it).

Reflecting on it, I think I was basically uncomfortable with the idea of adding a `pipeline` user to an ACL via something other than `CreatePipeline`, so what if we made `ModifyRoleBinding` return an error if `req.Principal` is not a user principal, and then _only_ `CreatePipeline` can add `pipeline` users to ACLs, via `CreatePipeline -> AddPipelineReaderToRepoInTransaction -> setUserRoleBindingInTransaction`?

If we do implement that, a corrolary idea is that maybe we should let `pach:root` circumvent this restriction and add anyone to an ACL (in case of Pachyderm bugs basically).

WDYT?

Reply:
> I was basically uncomfortable with the idea of adding a `pipeline` user to an ACL via something other than `CreatePipeline`

I don't think this is a security concern, right? It's mostly that we automatically manage pipeline permissions such that users _shouldn't_ need to change them. But I don't really forsee a huge downside to users manually adding/removing pipelines from repos where they have permissions to modify role bindings? The worst outcome I can see is a user breaks their own pipelines, and they can trivially fix it by updating the pipeline spec to fix the role bindings. 

One possible upside is being able to use the pipeline token for more than just reading inputs and writing output repos. It's possible users want to perform arbitrary RPCs against Pachyderm inside a pipeline, and I don't think we discourage that, but right now we make them generate a robot token manually and mount it in the pipeline. Is there a good reason they shouldn't be able to use the pipeline token for those operations, since it's already mounted and identified with the pipeline? In my mind robot tokens are mostly for external systems to talk to Pachyderm, and compute tasks inside Pachyderm use pipeline tokens.

I'm going to merge this so I can get all my changes in before PTO and they can get into the next alpha, but I'm open to discussing this more. Right now my instinct is that I want to avoid adding a special case (ModifyRoleBindings isn't allowed to work with pipeline principals) and then a special case to the special case (root user can modify role bindings for pipeline principals) but I'm not sure I fully understand the consequences of users modifying the role bindings manually.

Reply:
You make good points. Particularly the point about adding a pipeline to make `GetFile` work from inside the pipeline—I definitely see how adding the pipeline is cleaner than generating a robot token and having the pipeline use two tokens.

I guess the case I was thinking about is one where there's a pipeline I don't own and am not associated with. Could I change the behavior of that pipeline by creating a malicious repo and adding the pipeline to it? Maybe this is too contrived an example to worry about, but I guess what I'm wondering is, maybe it would make sense to say "I can add a pipeline to a repo if I can read the repo _and_ I can modify the pipeline?

Your point about not adding too many special cases also makes a ton of sense, though, and maybe the best option is to just wait until we've figured out pipeline permissions a bit more and see if a natural solution presents itself.

Reply:
> I guess the case I was thinking about is one where there's a pipeline I don't own and am not associated with. Could I change the behavior of that pipeline by creating a malicious repo and adding the pipeline to it? Maybe this is too contrived an example to worry about, but I guess what I'm wondering is, maybe it would make sense to say "I can add a pipeline to a repo if I can read the repo and I can modify the pipeline?

FWIW this is how GCP handles permissions - you can add any arbitrary service account to your resource role bindings, even across projects and organizations, because you're permitting them to make API calls but you can't force them to make those calls. So like, I could have a bucket full of malicious code and make it world-readable, but I can't force other people's GCE instances to download that code using only role bindings. 

I think the key distinction is that granting a pipeline user access to read from a repo is distinct from adding that repo to the pipeline spec.